httpee/http.go

```
@@ -0,0 +1,86 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*******************************************************************************/
+
+//Package httpee provides some convenience functions on top of the "http"
+//package from the stdlib.
+package httpee
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/sapcc/go-bits/logg"
+)
+
+// ContextWithSIGINT creates a new context.Context using the provided Context, and
+// launches a goroutine that cancels the Context when an interrupt signal is caught.
+func ContextWithSIGINT(ctx context.Context) context.Context {
+	ctx, cancel := context.WithCancel(ctx)
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-signalChan
+		signal.Reset(os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+		close(signalChan)
+		cancel()
+	}()
+	return ctx
+}
+
+// ListenAndServeContext is a wrapper around http.ListenAndServe() that additionally
+// shuts down the HTTP server gracefully when an error occurs or when an interrupt
+// signal is caught.
+func ListenAndServeContext(ctx context.Context, addr string, handler http.Handler) error {
+	server := &http.Server{Addr: addr, Handler: handler}
+
+	//waitForServerShutdown channel is used to block until server.Shutdown()
+	//has returned. This is because when Shutdown is called, Serve, ListenAndServe,
+	//and ListenAndServeTLS immediately return ErrServerClosed.
+	//We block to make sure the program doesn't exit prematurely.
+	waitForServerShutdown := make(chan struct{})
+	shutdownServer := make(chan int, 1)
+	go func() {
+		ctx = ContextWithSIGINT(ctx)
+		select {
+		case <-shutdownServer:
+			//continue down below
+		case <-ctx.Done():
+			logg.Error("Interrupt received...")
+		}
+
+		logg.Error("Shutting down HTTP server")
+		err := server.Shutdown(ctx)
+		if err != nil {
+			logg.Error("HTTP server shutdown: %s", err.Error())
```

I don't like that the error gets logged here directly, especially because this is a library function. Since `ListenAndServeContext` returns `error`, it should return this error instead. There will not be a conflict with the error returned by `ListenAndServe`:

1. Either `ListenAndServe` returns `ErrServerClosed`. In this case we return the error from `Shutdown` (which is possibly nil).
2. Or `ListenAndServe` returns some other error, then we return that. The error from `Shutdown` can be ignored in this case since the server is already dead anyway. (Or we could log something like `additional error encountered during shutdown: %s`.)